homeassistant/components/eafm/sensor.py

```diff
@@ -0,0 +1,188 @@
+
+    measurements = set()
+
+    async def async_update_data():
+        # DataUpdateCoordinator will handle aiohttp ClientErrors and timouts
+        async with async_timeout.timeout(30):
+            data = await get_station(session, station_key)
+
+        measures = get_measures(data)
+
+        # Look to see if payload contains new measures
+        for measure in measures:
+            if measure["@id"] not in measurements:
+                if "latestReading" not in measure:
+                    # Don't create a sensor entity for a gauge that isn't available
+                    continue
+
+                if measure["@id"] not in measurements:
+                    async_add_entities(
+                        [Measurement(hass.data[DOMAIN][station_key], measure["@id"])]
+                    )
+                    measurements.add(measure["@id"])
+
+        # Turn data.measures into a dict rather than a list so easier for entities to
+        # find themselves.
+        data["measures"] = {measure["@id"]: measure for measure in measures}
+
+        return data
+
+    hass.data[DOMAIN][station_key] = coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="sensor",
+        update_method=async_update_data,
+        update_interval=timedelta(seconds=15 * 60),
+    )
+
+    # Fetch initial data so we have data when entities subscribe
+    await coordinator.async_refresh()
+
+    return len(measurements) > 0
+
+
+class Measurement(Entity):
+    """A gauge at a flood monitoring station."""
+
+    def __init__(self, coordinator, key):
+        """Initialise the gauge with a data instance and station."""
+        self.coordinator = coordinator
+        self.key = key
+
+    @property
+    def station_name(self):
+        """Return the station name for the measure."""
+        return self.coordinator.data["label"]
+
+    @property
+    def station_id(self):
+        """Return the station id for the measure."""
+        return self.coordinator.data["stationReference"]
+
+    @property
+    def qualifier(self):
+        """Return the qualifier for the station."""
+        return self.coordinator.data["measures"][self.key]["qualifier"]
+
+    @property
+    def parameter_name(self):
+        """Return the parameter name for the station."""
+        return self.coordinator.data["measures"][self.key]["parameterName"]
+
+    @property
+    def name(self):
+        """Return the name of the gauge."""
+        return f"{self.station_name} {self.parameter_name} {self.qualifier}"
+
+    @property
+    def should_poll(self) -> bool:
+        """Stations are polled as a group - the entity shouldn't poll by itself."""
+        return False
+
+    @property
+    def unique_id(self):
+        """Return the unique id of the gauge."""
+        return self.key
+
+    @property
+    def device_info(self):
+        """Return the device info."""
+        return {
+            "identifiers": {(DOMAIN, "measure-id", self.unique_id)},
```

Comment:
Is every entity a new service?

Reply:
You are right. Each sensor represents a physical device at a river somewhere, but they should be grouped by the station reference. That's implemented in the latest commit.